File: 2024/Level 24/level_24_part_b.py
import os

def read_input(file_path):
    """
    Reads the input file and splits it into wires and gates sections.
    
    :param file_path: Path to the input file.
    :return: Tuple containing lists of wires and gates.
    """
    with open(file_path, 'r', encoding='utf-8') as file:
        content = file.read().strip()
        wires_raw, gates_raw = content.split("\n\n")
        return wires_raw.split("\n"), gates_raw.split("\n")

def parse_wires(wires_raw):
    """
    Parses the raw wires data into a dictionary.
    
    :param wires_raw: List of wire definitions in 'name: value' format.
    :return: Dictionary mapping wire names to their integer values.
    """
    wires = {}
    for line in wires_raw:
        name, value = line.split(": ")
        wires[name] = int(value)
    return wires

def parse_gates(gates_raw, wires):
    """
    Parses the raw gates data into a list of gate dictionaries.
    
    :param gates_raw: List of gate definitions in 'inputs -> output' format.
    :param wires: Dictionary of existing wires to initialize new ones as None.
    :return: List of gate dictionaries with keys 'a', 'op', 'b', and 'output'.
    """
    gates = []
    for line in gates_raw:
        inputs, output = line.split(" -> ")
        parts = inputs.split(" ")
        
        if len(parts) == 3:
            a, op, b = parts
        else:
            a, op = parts
            b = None

        gate = {'a': a, 'op': op, 'b': b, 'output': output}
        gates.append(gate)

        # Initialize wires if they don't exist
        for wire in [a, b, output]:
            if wire and wire not in wires:
                wires[wire] = None

    return gates

def is_direct(gate):
    """Checks if either input starts with 'x'."""
    return gate['a'].startswith('x') or (gate['b'] and gate['b'].startswith('x'))

def is_output(gate):
    """Checks if the output starts with 'z'."""
    return gate['output'].startswith('z')

def is_gate(type_op):
    """Returns a function that checks if a gate's operation matches the specified type."""
    return lambda gate: gate['op'] == type_op

def has_output(output):
    """Returns a function that checks if a gate's output matches the specified output."""
    return lambda gate: gate['output'] == output

def has_input(input_wire):
    """Returns a function that checks if a gate's inputs include the specified wire."""
    return lambda gate: gate['a'] == input_wire or gate['b'] == input_wire

def main():
    """
    Main function to read input, parse data, and compute flagged outputs.
    """
    # Read and parse input
    wires_raw, gates_raw = read_input("input_level_24.txt")
    wires = parse_wires(wires_raw)
    input_bit_count = len(wires_raw) // 2
    gates = parse_gates(gates_raw, wires)
    
    # Initialize set for flagged outputs
    flags = set()

    # Filter gates into categories
    FA_GATE0 = list(filter(is_direct, filter(is_gate("XOR"), gates)))
    FA_GATE3 = list(filter(lambda g: g['op'] == "XOR" and not is_direct(g), gates))
    output_gates = list(filter(is_output, gates))
    
    # Check FA Gate0 gates
    for gate in FA_GATE0:
        a, b, output = gate['a'], gate['b'], gate['output']
        is_first = a == "x00" or b == "x00"
        if is_first:
            if output != "z00":
                flags.add(output)
        elif output == "z00":
            flags.add(output)
        elif is_output(gate):
            flags.add(output)

    # Check FA Gate3 gates
    for gate in FA_GATE3:
        if not is_output(gate):
            flags.add(gate['output'])

    # Check output gates
    for gate in output_gates:
        is_last = gate['output'] == f"z{str(input_bit_count).zfill(2)}"
        if is_last:
            if gate['op'] != "OR":
                flags.add(gate['output'])
        elif gate['op'] != "XOR":
            flags.add(gate['output'])

    # Additional checks for FA Gate0
    check_next = []
    for gate in FA_GATE0:
        output = gate['output']
        if output in flags or output == "z00":
            continue
        
        matches = list(filter(has_input(output), FA_GATE3))
        if not matches:
            check_next.append(gate)
            flags.add(output)

    for gate in check_next:
        a, output = gate['a'], gate['output']
        intended_result = f"z{a[1:]}"
        matches = list(filter(has_output(intended_result), FA_GATE3))
        if len(matches) != 1:
            raise Exception("Wrong input.")
        
        match = matches[0]
        to_check = [match['a'], match['b']]
        or_matches = list(filter(is_gate("OR"), gates))
        or_matches = list(filter(lambda g: g['output'] in to_check, or_matches))
        
        if len(or_matches) != 1:
            raise Exception("Too complex.")
        
        or_match_output = or_matches[0]['output']
        correct_output = next((wire for wire in to_check if wire != or_match_output), None)
        flags.add(correct_output)

    # Final validation
    if len(flags) != 8:
        raise Exception("Too complex.")
    
    print(",".join(sorted(flags)))

if __name__ == "__main__":
    main()

File: 2024/Level 24/level24_part_b.lua
local function read_input(file_path)
    local file = io.open(file_path, "r")
    if not file then
        error("Failed to open file: " .. file_path)
    end
    local content = file:read("*all")
    file:close()
    
    -- Normalize line endings and split into lines
    local lines = {}
    for line in content:gsub("\r\n", "\n"):gsub("\r", "\n"):gmatch("[^\n]+") do
        table.insert(lines, line)
    end
    
    -- Split into wires and gates based on content
    local wires_raw = {}
    local gates_raw = {}
    local is_gate_section = false
    
    for _, line in ipairs(lines) do
        if line:find("->") then  -- Gate line detection
            is_gate_section = true
            table.insert(gates_raw, line)
        elseif not is_gate_section then  -- Wire line
            table.insert(wires_raw, line)
        else
            error("Mixed content format")
        end
    end

    return wires_raw, gates_raw
end

local function parse_wires(wires_raw)
    local wires = {}
    for _, line in ipairs(wires_raw) do
        local name, value = line:match("([^:]+):%s*(%d+)")
        if name and value then
            wires[name] = tonumber(value)
        else
            error("Invalid wire format: " .. line)
        end
    end
    return wires
end

local function parse_gates(gates_raw)
    local gates = {}
    for _, line in ipairs(gates_raw) do
        local inputs, output = line:match("([^->]+) %-> ([^%s]+)")
        if not inputs or not output then
            error("Invalid gate format: " .. line)
        end
        local parts = {}
        for part in inputs:gmatch("%S+") do
            table.insert(parts, part)
        end
        local a, op, b
        if #parts == 3 then
            a, op, b = parts[1], parts[2], parts[3]
        elseif #parts == 2 then
            a, op = parts[1], parts[2]
            b = nil
        else
            error("Invalid gate inputs: " .. inputs)
        end
        local gate = { a = a, op = op, b = b, output = output }
        table.insert(gates, gate)
    end
    return gates
end

local function main()
    local wires_raw, gates_raw = read_input("input_level_24.txt")
    local wires = parse_wires(wires_raw)
    local input_bit_count = math.floor(#wires_raw / 2)
    local gates = parse_gates(gates_raw)

    local flags = {}

    -- Categorize gates
    local FA_GATE0 = {}
    local FA_GATE3 = {}
    for _, gate in ipairs(gates) do
        if gate.op == "XOR" then
            local a_start = gate.a:sub(1, 1) == 'x'
            local b_start = gate.b and gate.b:sub(1, 1) == 'x'
            if a_start or b_start then
                table.insert(FA_GATE0, gate)
            else
                table.insert(FA_GATE3, gate)
            end
        end
    end

    local output_gates = {}
    for _, gate in ipairs(gates) do
        if gate.output:sub(1, 1) == 'z' then
            table.insert(output_gates, gate)
        end
    end

    -- Check FA_GATE0 gates
    for _, gate in ipairs(FA_GATE0) do
        local a = gate.a
        local b = gate.b
        local output = gate.output
        local is_first = (a == "x00") or (b == "x00")
        if is_first then
            if output ~= "z00" then
                flags[output] = true
            end
        else
            if output == "z00" then
                flags[output] = true
            elseif output:sub(1, 1) == 'z' then
                flags[output] = true
            end
        end
    end

    -- Check FA_GATE3 gates
    for _, gate in ipairs(FA_GATE3) do
        if gate.output:sub(1, 1) ~= 'z' then
            flags[gate.output] = true
        end
    end

    -- Check output gates
    for _, gate in ipairs(output_gates) do
        local output = gate.output
        local is_last = output == string.format("z%02d", input_bit_count)
        if is_last then
            if gate.op ~= "OR" then
                flags[output] = true
            end
        else
            if gate.op ~= "XOR" then
                flags[output] = true
            end
        end
    end

    -- Additional checks for FA_GATE0
    local check_next = {}
    for _, gate in ipairs(FA_GATE0) do
        local output = gate.output
        if not flags[output] and output ~= "z00" then
            local matches = {}
            for _, g3 in ipairs(FA_GATE3) do
                if g3.a == output or g3.b == output then
                    table.insert(matches, g3)
                end
            end
            if #matches == 0 then
                table.insert(check_next, gate)
                flags[output] = true
            end
        end
    end

    for _, gate in ipairs(check_next) do
        local a = gate.a
        local output = gate.output
        local num = a:sub(2)
        local intended_result = "z" .. num
        local matches = {}
        for _, g3 in ipairs(FA_GATE3) do
            if g3.output == intended_result then
                table.insert(matches, g3)
            end
        end
        if #matches ~= 1 then
            error("Wrong input.")
        end
        local match = matches[1]
        local to_check = { match.a, match.b }
        local or_matches = {}
        for _, g in ipairs(gates) do
            if g.op == "OR" and (g.output == to_check[1] or g.output == to_check[2]) then
                table.insert(or_matches, g)
            end
        end
        if #or_matches ~= 1 then
            error("Too complex.")
        end
        local or_output = or_matches[1].output
        local correct_output
        for _, wire in ipairs(to_check) do
            if wire ~= or_output then
                correct_output = wire
                break
            end
        end
        flags[correct_output] = true
    end

    -- Validate and print
    local count = 0
    for _ in pairs(flags) do
        count = count + 1
    end
    if count ~= 8 then
        error("Too complex.")
    end

    local sorted_flags = {}
    for k in pairs(flags) do
        table.insert(sorted_flags, k)
    end
    table.sort(sorted_flags)
    print(table.concat(sorted_flags, ","))
end

main()
